cmd/monorail/src/lib.rs

```diff
@@ -0,0 +1,731 @@
+    fn requires_target(&self) -> bool {
+        !matches!(
+            &self.cmd,
+            Command::Info { .. }
+                | Command::Phy { cmd: PhyCommand::Info { .. }, .. }
+        )
+    }
+}
+
+#[derive(Parser, Debug)]
+enum Command {
+    Info {
+        reg: String,
+        #[clap(parse(try_from_str = parse_int::parse))]
+        value: Option<u32>,
+    },
+    Status {
+        /// arguments
+        #[clap(long, short, use_value_delimiter = true)]
+        ports: Vec<u8>,
+    },
+    Counters {
+        #[clap(long, short)]
+        port: u8,
+        #[clap(long, short)]
+        reset: bool,
+    },
+    Read {
+        reg: String,
+    },
+    Write {
+        reg: String,
+        #[clap(parse(try_from_str = parse_int::parse))]
+        value: u32,
+    },
+    Phy {
+        #[clap(subcommand)]
+        cmd: PhyCommand,
+    },
+}
+
+#[derive(Parser, Debug)]
+enum PhyCommand {
+    Info {
+        reg: String,
+    },
+    Read {
+        #[clap(long, short)]
+        port: u8,
+        reg: String,
+    },
+    Write {
+        #[clap(long, short)]
+        port: u8,
+        reg: String,
+        #[clap(parse(try_from_str = parse_int::parse))]
+        value: u16,
+    },
+}
+
+/// Parses either a register address (as an integer or hex literal), or a
+/// string representing a register's name.
+fn parse_reg_or_addr(s: &str) -> Result<TargetRegister> {
+    let reg = if let Ok(addr) = parse_int::parse(s) {
+        TargetRegister::from_addr(addr)?
+    } else {
+        s.parse()?
+    };
+    Ok(reg)
+}
+
+fn pretty_print_fields(value: u32, fields: &BTreeMap<String, Field<String>>) {
+    let mut field_keys = fields.keys().collect::<Vec<_>>();
+    if field_keys.is_empty() {
+        return;
+    }
+    field_keys.sort_by(|a, b| fields[*b].lo.cmp(&fields[*a].lo));
+    println!("  bits |    value   | field");
+    for f in field_keys {
+        let field = &fields[f];
+        let bits = (value & ((1u64 << field.hi) - 1) as u32) >> field.lo;
+        println!(" {:>2}:{:<2} | 0x{:<8x} | {}", field.hi, field.lo, bits, f);
+    }
+}
+
+fn monorail_read(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    context: &mut HiffyContext,
+    reg: String,
+) -> Result<()> {
+    let reg = parse_reg_or_addr(&reg)?;
+    let addr = reg.address();
+    humility::msg!("Reading {} from 0x{:x}", reg, addr);
+
+    let op = IdolOperation::new(hubris, "Monorail", "read_vsc7448_reg", None)?;
+    let value = humility_cmd_hiffy::hiffy_call(
+        hubris,
+        core,
+        context,
+        &op,
+        &[("addr", IdolArgument::Scalar(u64::from(addr)))],
+    )?;
+    match value {
+        Ok(v) => {
+            let value = if let Value::Base(Base::U32(v)) = v {
+                v
+            } else {
+                bail!("Got bad reflected value: expected U32, got {:?}", v);
+            };
+            println!("{} => 0x{:x}", reg, value);
+            if value == 0x88888888 {
```

Comment:
Where does this value come from?

Reply:
It's built into the VSC7448's SPI mode; added a comment to explain.